lightning/src/routing/router.rs

```diff
@@ -876,40 +878,126 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+				// We start building the path from reverse, i.e., from payee 
+				// to the first RouteHintHop in the path.
+				let route_it = route.iter().rev();
+				let mut prev_route_it = route.iter().rev();
+				let mut aggregate_next_hops_fee_msat = 0;
+				let mut aggregate_next_hops_path_htlc_minimum_msat = 0;
+				let mut hop_used = true;
+				let mut prev_hop;
+
+				for (idx, hop) in route_it.enumerate(){
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+					// really sucks, cause we're gonna need that eventually.
+					let hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+						Some(htlc_minimum_msat) => htlc_minimum_msat,
+						None => 0
+					};
+
+					let directional_info = DummyDirectionalChannelInfo {
+						cltv_expiry_delta: hop.cltv_expiry_delta as u32,
+						htlc_minimum_msat: hop_htlc_minimum_msat,
+						htlc_maximum_msat: hop.htlc_maximum_msat,
+						fees: hop.fees,
+					};
+
+					prev_hop = match idx {
+						0 => &payee,
+						_ => match prev_route_it.next() {
+							Some(x) => &x.src_node_id,
+							None => &payee,
+						}
+					};
+
+					if idx == 0 {
+						// We assume that the recipient only included route hints for routes which had
+						// sufficient value to route `final_value_msat`. Note that in the case of "0-value"
+						// invoices where the invoice does not specify value this may not be the case, but
+						// better to include the hints than not.
+						if add_entry!(hop.short_channel_id, hop.src_node_id, payee, directional_info, Some((final_value_msat + 999) / 1000), &empty_channel_features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat) {
+							aggregate_next_hops_fee_msat = match aggregate_next_hops_fee_msat.checked_add(final_value_msat) {
+								Some(value) => value,
+								None => u64::max_value()
+							};
+						} else {
+							hop_used = false;
+						};
+					} else {
+						// If this is not the last hop then check if it is a feasible intermediate hop
+						// otherwise test the previous hop for direct channel path and break.
+						let reqd_channel_cap = Some(match aggregate_next_hops_fee_msat.checked_add(999) {
+							Some(cap) => cap,
+							None => u64::max_value()
+						} / 1000); // converting from msat
+
+						if !add_entry!(hop.short_channel_id, hop.src_node_id, prev_hop, directional_info, reqd_channel_cap, &empty_channel_features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat) {
+							hop_used = false;
+						}
+					}
+
+					if !hop_used {
+						// If this hop was not used then there is no use checking the preceeding hops
+						// in the RouteHint. We can break by just searching for a direct channel between 
+						// last checked hop and first_hop_targets
+						if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&prev_hop) {
```

Comment:
Why is this inside the `hop_used` branch? If we just move it outside of the if we can drop the `idx == route.len() - 1` if as well, no? I mean we could probably also drop this hunk entirely and it wouldn't matter, but its simpler if we just move it out of the if.

Reply:
The reason is if the current hop being assessed was not used then the direct channel needs to be created from the previously added hop. If the current hop is the last hop then that would mean from payee to the ```first_hop_targets```. However, if ```idx == route.len() - 1``` was triggered then this hop was used and in this case the direct channel will be from this hop to first_hop_targets.

Reply:
If a particular hop in a ```RouteHint``` is not used then the loop is broken. But ```idx == route.len() - 1``` is only triggered if all the ```RouteHintHop```s are used.

Reply:
Hmm, right. Its still a bit simpler if we move the pre-`break` `add_entry!` call out of the `else` block and just always do it, though.